### DISCUSSION TITLE
Composing two components programmatically

### DISCUSSION THREAD

**Discussion by taybenlor** [bytecodealliance/wasm-tools#1443] [Q&A] [answered]
Hi!

I want to use the `ComponentComposer` from Rust to compose a WASM component I'm generating with an existing component on the filesystem. I'm finding the interface to composition quite convoluted for this usecase.

I tried looking through the source to see if there's a nicer way, but I'm finding it hard to follow.

Is there a simple way to take two components and compose them together?

Cheers,
Ben

**Comment by alexcrichton** [comment]
I think you've found what the `wasm-compose` crate has to offer, so if you're finding that difficult to use that crate and/or this repo may not have much more to offer. You can try taking a look at the [source for the `compose` subcommand](https://github.com/bytecodealliance/wasm-tools/blob/main/src/bin/wasm-tools/compose.rs), but that may not help more than you've already seen.

The original author of `wasm-compose`, however, has been working with other folks on a [new tool](https://github.com/peterhuene/wac) to make composition easier. You might be interested in giving that a spin and seeing if it meets your use case?

**Reply by taybenlor** [reply]
Thanks heaps! Super helpful answer. It's hard to navigate all this tooling (I realise it's all very fresh) so pointers are really appreciated.

Have a good day!

**Reply by peterhuene** [reply, marked as answer]
Hi @taybenlor. I'm happy to help with the `wasm-compose` API.

The [`ComponentComposer`](https://docs.rs/wasm-compose/latest/wasm_compose/composer/struct.ComponentComposer.html) is simplistic and is essentially a wrapper around [`CompositionGraph`](https://docs.rs/wasm-compose/latest/wasm_compose/graph/struct.CompositionGraph.html) (which is the *flexible* primitive for composing components together) geared towards the CLI experience; it expects a configuration for figuring out how to compose components together and that everything to compose exists on disk.

With the `CompositionGraph`, however, you can compose together components either defined entirely in memory or read from disk. The basic concept for this API is that you "add" components to the graph so that the graph is aware of them, and then create instances (nodes in the graph) and connect (i.e. create an edge) between an instance's export and another instance's imports. You then encode the graph into a new component.

The `wac` tool that Alex mentioned is a work-in-progress; it has a much more flexible language for expressing compositions than the configuration file of the current `wasm-tools compose` tool, but it's entirely lacking a programmatic API (something we have an issue for and will likely implement something quite similiar to the existing `CompositionGraph` API).

I hope that helps some and feel free to reach out to me on GitHub or the Bytecode Alliance zulip.

**Reply by taybenlor** [reply]
Oh fantastic, thanks @peterhuene! I'll have a go with CompositionGraph and then reach out if I get stuck.
